src/events/eventsource.ts

```diff
@@ -0,0 +1,267 @@
+    super();
+    this.readyState = ReadyState.CONNECTING;
+    const req = got.stream(this.url, { isStream: true });
+    this.req = req;
+    req.on('response', (res: Response) => {
+      if (res.statusCode === 500 || res.statusCode === 502 || res.statusCode === 503 || res.statusCode === 504) {
+        this.emit('error', newErrorEvent(res.statusMessage, res.statusCode));
+        this.onConnectionClosed();
+        return;
+      }
+
+      // TODO: should we handle redirects?
+
+      if (res.statusCode !== 200) {
+        this.emit('error', newErrorEvent(res.statusMessage, res.statusCode));
+        this.close();
+        return;
+      }
+
+      this.readyState = ReadyState.OPEN;
+
+      req.on('end', () => {
+        req.removeAllListeners('close');
+        req.removeAllListeners('end');
+        this.onConnectionClosed('Stream ended');
+      });
+
+      req.on('close', () => {
+        req.removeAllListeners('close');
+        req.removeAllListeners('end');
+        this.onConnectionClosed('Stream closed');
+      });
+
+      this.emit('open', { type: 'open' });
+    });
+
+    let buf: Buffer;
+    let newBuffer: Buffer;
+    let startingPos = 0;
+    let startingFieldLength = -1;
+    let newBufferSize = 0;
+    let bytesUsed = 0;
+    let discardTrailingNewline = false;
+
+    req.on('data', chunk => {
+      if (!buf) {
+        buf = chunk;
+        if (hasBom(buf)) {
+          buf = buf.subarray(bom.length);
+        }
+        bytesUsed = buf.length;
+      } else {
+        if (chunk.length > buf.length - bytesUsed) {
+          newBufferSize = buf.length * 2 + chunk.length;
+          if (newBufferSize > maxBufferAheadAllocation) {
+            newBufferSize = buf.length + chunk.length + maxBufferAheadAllocation;
+          }
+          newBuffer = Buffer.alloc(newBufferSize);
+          buf.copy(newBuffer, 0, 0, bytesUsed);
+          buf = newBuffer;
+        }
+        chunk.copy(buf, bytesUsed);
+        bytesUsed += chunk.length;
+      }
+
+      let pos = 0;
+      const length = bytesUsed;
+
+      while (pos < length) {
+        if (discardTrailingNewline) {
+          if (buf[pos] === lineFeed) {
+            ++pos;
+          }
+          discardTrailingNewline = false;
+        }
+
+        let lineLength = -1;
+        let fieldLength = startingFieldLength;
+        let c: number;
+
+        for (let i = startingPos; lineLength < 0 && i < length; ++i) {
+          c = buf[i];
+          if (c === colon) {
+            if (fieldLength < 0) {
+              fieldLength = i - pos;
+            }
+          } else if (c === carriageReturn) {
+            discardTrailingNewline = true;
+            lineLength = i - pos;
+          } else if (c === lineFeed) {
+            lineLength = i - pos;
+          }
+        }
+
+        if (lineLength < 0) {
+          startingPos = length - pos;
+          startingFieldLength = fieldLength;
```

Comment:
in this case it is a `field`, right?

can we therefore parse this as:

`this.parseEventStreamField(buf, pos, fieldLength)` ?

And the other case as:

`this.parseEventStreamLine(buf, pos, lineLength);` ?

If seems now to rely on a singe function with another check. This looks more complex than it needs to be.

Reply:
That is for case when we don't receive 'full' line, as it is not guaranty that we will have whole line
so we just make sure that we have a complete line before parse it.